Add Instructor Buttons for assessments from student detail page

nicolasruscher opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Currently, when assessing an exam, instructors will take a look at specific student exams. In the current view, they can see the points of the most recent result, but the cannot open the most recent result. In this PR we change this. 

### Description
<!-- Describe your changes in detail -->
We now introduce new buttons for instructors to access the results by their id. Within the student exam view instructors can directly open up the most recent result. As we do not know which correction round it belongs to, we added new routes that work by resultId. For Quiz Exercises no link is displayed. Programming Exercises have a link to the Submission Dashboard as they can have only automatic results enabled. 

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Create an Exam and submit some student exams. 
4. Test that instructors can open up assessment over their respective submissions dashboard for all correction rounds as usual. 
5. Test that whenever assessing the submissions, instructors can go to StudentExams -> Specific Studentexam -> when clicking on the Button to open up the latest Assessment, they are shown the latest assessment. 
6. Test that Tutor and instructors can assess submissions / open assessments for all correction rounds within the exercise dashboard as usual. 
7. Test that normal exercise assessments can be opened as usual via the exercise dashboards

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
student-exam-detail: 86.41
TextAssessmentResource: 86%
ModelingSubmissionResource: 95%
FileUploadAssessmentResource: 98%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![Bildschirmfoto 2021-03-15 um 23 14 48](https://user-images.githubusercontent.com/42205689/111228339-529d2480-85e4-11eb-9a05-bc99fcacfeeb.png)

derLalla commented on src/main/webapp/app/entities/submission.model.ts:
```diff
@@ -76,6 +76,20 @@ export function getSubmissionResultByCorrectionRound(submission: Submission | un
+ * @returns the results or undefined if submission or the result for the requested id is undefined
+ */
+export function getSubmissionResultById(submission: Submission | undefined, resultId: number): Result | undefined {
+    if (submission && submission.results) {
+        return submission.results.find((result) => result.id === resultId);
+    }
+    return undefined;
+}
```
you can simplify it with this:

```suggestion
    return submission?.results?.find((result) => result.id === resultId);
}
```

derLalla commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -147,8 +153,16 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
-                            <ng-container *ngIf="exercise?.studentParticipations && exercise?.studentParticipations.length > 0 && (exercise?.studentParticipations)![0].results">
-                                {{ rounding(((exercise?.studentParticipations)![0].results![0].score * exercise?.maxPoints) / 100) }}
+                            <ng-container
+                                *ngIf="
+                                    exercise?.studentParticipations &&
+                                    exercise?.studentParticipations.length > 0 &&
+                                    exercise?.studentParticipations?.length > 0 &&
+                                    (exercise?.studentParticipations)![0].submissions?.length > 0 &&
+                                    (exercise?.studentParticipations)![0].submissions[0].results
+                                "
```
usually this would be enough, but the linter gives an error: `if exercise?.studentParticipations?.[0]?.submissions?.[0]?.results`

```suggestion
                                *ngIf="
                                    exercise?.studentParticipations?.length > 0 &&
                                    exercise.studentParticipations![0].submissions?.length > 0 &&
                                    exercise.studentParticipations![0].submissions[0].results
                                "
```

derLalla commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -162,15 +176,68 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
+                                    exercise.type != ExerciseType.PROGRAMMING
+                                "
+                            >
+                                <div *ngIf="(exercise?.studentParticipations)![0].results?.length > 0">
```
you don't need to use brackets, in all cases: 

```suggestion
                                <div *ngIf="exercise?.studentParticipations![0].results?.length > 0">
```

derLalla reviewed (changes requested):
I have left some comments in the code.

FuchsDominik commented on src/main/webapp/app/exercises/text/assess/text-assessment.service.ts:
```diff
@@ -267,4 +273,20 @@ export class TextAssessmentService {
         }
     }
+
+    /**
```
We should probably also extend the description and not only state params which we can also see in the method as well

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,22 +147,33 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
         User user = userRepository.getUserWithGroupsAndAuthorities();
-        if (!authCheckService.isAtLeastTeachingAssistantForExercise(fileUploadExercise, user)) {
+        if (!authCheckService.isAtLeastTeachingAssistantForExercise(fileUploadExercise, user)
+                || resultId > 0 && !authCheckService.isAtLeastInstructorForExercise(fileUploadExercise, user)) {
```
This should probably change to `!= 0` to match the check below.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,22 +147,33 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
+
+        fileUploadExercise.setGradingCriteria(gradingCriteria);
+
+        if (resultId != 0 && authCheckService.isAtLeastInstructorForExercise(fileUploadExercise, user)) {
```
We can drop the auth service check here since the case where the correction round is non-zero and we are not a tutor is already covered above. So this simplifies to `if (resultId != 0)`

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java:
```diff
@@ -191,9 +192,17 @@ public ModelingSubmissionResource(SubmissionRepository submissionRepository, Res
             return forbidden();
         }
+
+        if (resultId != 0 && authCheckService.isAtLeastInstructorForExercise(modelingExercise, user)) {
```
Same here.

FuchsDominik reviewed (approved):
Looks good to me :)

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:
```diff
@@ -270,12 +273,23 @@ public TextAssessmentResource(AuthorizationCheckService authCheckService, TextAs
         checkAuthorization(exercise, user);
         final boolean isAtLeastInstructorForExercise = authCheckService.isAtLeastInstructorForExercise(exercise, user);
 
+        if (!authCheckService.isAtLeastTeachingAssistantForExercise(exercise, user) || resultId > 0 && !authCheckService.isAtLeastInstructorForExercise(exercise, user)) {
```
This can use `isAtLeastInstructorForExercise` instead of calling the authCheckService for the last part of the if.

FrankeLukas commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -26,7 +26,7 @@ import { StructuredGradingCriterionService } from 'app/exercises/shared/structur
 import { ExerciseType, getCourseFromExercise } from 'app/entities/exercise.model';
 import { Authority } from 'app/shared/constants/authority.constants';
-import { getLatestSubmissionResult } from 'app/entities/submission.model';
+import { getLatestSubmissionResult, getSubmissionResultByCorrectionRound, getSubmissionResultById } from 'app/entities/submission.model';
```
`getSubmissionResultByCorrectionRound` is unused.

ivanchimeno commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -54,6 +60,7 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
     noNewSubmissions: boolean;
     hasAssessmentDueDatePassed: boolean;
     correctionRound: number;
+    resultId?: number;
```
Yo don't need to use `undefined`, since you're always checking for `resultId != 0` you can just initialize it to 0.

ivanchimeno replied:
```suggestion
    resultId: number;
```

ivanchimeno commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -110,6 +117,7 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
         this.unusedTextBlockRefs = [];
         this.complaint = undefined;
         this.totalScore = 0;
+        this.resultId = undefined;
```
```suggestion
        this.resultId = 0;
```

ivanchimeno commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -134,6 +142,7 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
 
         this.activatedRoute.paramMap.subscribe((paramMap) => {
             this.exerciseId = Number(paramMap.get('exerciseId'));
+            this.resultId = Number(paramMap.get('resultId'));
```
```suggestion
            this.resultId = Number(paramMap.get('resultId')) ?? 0;
```

ivanchimeno commented on src/main/webapp/app/exercises/text/assess/text-submission-assessment.component.ts:
```diff
@@ -160,7 +169,13 @@ export class TextSubmissionAssessmentComponent extends TextAssessmentBaseCompone
         setLatestSubmissionResult(this.submission, getLatestSubmissionResult(this.submission));
-        this.result = getSubmissionResultByCorrectionRound(this.submission, this.correctionRound);
+
+        if (this.resultId && this.resultId !== 0) {
```
```suggestion
        if (this.resultId > 0) {
```

nicolasruscher replied:
done

ivanchimeno commented on src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts:
```diff
@@ -63,6 +63,7 @@ export class ModelingAssessmentEditorComponent implements OnInit {
     hasAutomaticFeedback = false;
     hasAssessmentDueDatePassed: boolean;
     correctionRound = 0;
+    resultId?: number;
```
Same suggestions would apply (see below)

nicolasruscher replied:
done

ivohashamov reviewed (dismissed):
Code looks good

SimonEntholzer commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:
```diff
@@ -209,4 +220,34 @@ export class StudentExamDetailComponent implements OnInit {
+        } else if (submission) {
+            this.openingAssessmentEditorForNewSubmission = true;
+            route = getLinkToSubmissionAssessment(exercise.type, this.courseId, exercise.id!, submission.id!, this.examId, exercise.exerciseGroup?.id!, resultId);
+            /*if (resultId) {
```
I guess this comment can be removed?

nicolasruscher replied:
yes, you are right

SimonEntholzer commented on src/main/webapp/app/exercises/file-upload/participate/file-upload-submission.service.ts:
```diff
@@ -36,11 +36,16 @@ export class FileUploadSubmissionService {
+            // in case resultId is set, we do not need the correction round
+            params = params.set('resultId', resultId!.toString());
+        } else {
             params = params.set('correction-round', correctionRound.toString());
```
Is there a reason we should not send the correction-round, if we send the result Id? I get that we theoretically don't need it, but I don't know for sure how the server handles this. Maybe there could be a conflict when the correctionRound in the client is 1, we only send the resultId, and the server, by default assumes the correctionRound is 0.

nicolasruscher replied:
In the case where we have the resultId we don't know the correction round, that's why we use the resultId instead.

SimonEntholzer reviewed (changes requested):
Good changes :D I really like that we use the resultId in the assessments now. That will probably remove a lot of ambiguity. 
I just left a minor comment, no sure if this should be adapted or not though.

SimonEntholzer reviewed (dismissed):
Code looks good, tested on TS3, and everything works fine.
One thing I noticed: the "assessment Dashboard" button leads to the "assessment Dashborad" which is usually accessed via the button "Submissions". Theoretically you should rename it for consistency reasons, but I guess like this it is fine too.
<details>
<summary>Screenshots</summary>
<img src="https://user-images.githubusercontent.com/33342534/111622021-fc4d0300-87e8-11eb-8931-2a780e8acb9a.png">
<img src="https://user-images.githubusercontent.com/33342534/111622306-4afa9d00-87e9-11eb-8cd7-5f4ba7090795.png">
</details>

SimonEntholzer reviewed (approved):
Re-approve

FuchsDominik reviewed (approved):
Tested and all testing steps worked without any encountered problems :)

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,29 +147,46 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
      *
      * @param submissionId the id of the fileUploadSubmission to retrieve
      * @param correctionRound the correctionRound of the result we want to receive
+     * @param resultId the resultId for which we want do get the submission
```
```suggestion
     * @param resultId the resultId for which we want to get the submission
```

nicolasruscher replied:
Done

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,29 +147,46 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
         var fileUploadSubmission = fileUploadSubmissionService.findOne(submissionId);
         var studentParticipation = (StudentParticipation) fileUploadSubmission.getParticipation();
         var fileUploadExercise = (FileUploadExercise) studentParticipation.getExercise();
         var gradingCriteria = gradingCriterionService.findByExerciseIdWithEagerGradingCriteria(fileUploadExercise.getId());
```
I guess we can move this line down as well

nicolasruscher replied:
Done

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,29 +147,46 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
     @PreAuthorize("hasAnyRole('TA','INSTRUCTOR','ADMIN')")
     public ResponseEntity<FileUploadSubmission> getFileUploadSubmission(@PathVariable Long submissionId,
-            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound) {
+            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound, @RequestParam(value = "resultId", defaultValue = "0") long resultId) {
```
Would it not make sense in this case, to use `Long resultId` with a default value of `null`? (this default value might be implicitly set by the way)

krusche replied:
```suggestion
            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound, @RequestParam(value = "resultId") Long resultId) {
```

nicolasruscher replied:
Done

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,29 +147,46 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
+
+        fileUploadExercise.setGradingCriteria(gradingCriteria);
+
+        if (resultId > 0) {
```
Please add some discussion about the logic of this REST call to the JavaDoc above. In particular that the resultId is used first. In case it is not specified, the correction round is used. In case the correction round is not specified, the first correction round is assumed

krusche replied:
```suggestion
        if (resultId != null) {
```

nicolasruscher replied:
Done

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,29 +147,46 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
         this.fileUploadSubmissionService.hideDetails(fileUploadSubmission, user);
 
-        if (correctionRound == 0 && fileUploadSubmission.getResults().size() == 2) {
+        if (correctionRound == 0 && resultId == 0 && fileUploadSubmission.getResults().size() == 2) {
```
in the unlikely case we will have more than 2 results in the future (for whatever reason), would it not make sense to leave out the last part here?

krusche replied:
```suggestion
        if (correctionRound == 0 && resultId == null) {
```

nicolasruscher replied:
I changed the check and included it into the other calls.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java:
```diff
@@ -176,13 +176,14 @@ public ModelingSubmissionResource(SubmissionRepository submissionRepository, Res
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<ModelingSubmission> getModelingSubmission(@PathVariable Long submissionId,
-            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound) {
+            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound, @RequestParam(value = "resultId", defaultValue = "0") long resultId) {
```
Same feedback as for the other method for file upload submissions.
Please improve JavaDoc, use Long resultId and double check why the logic regarding the number of results is missing here?

nicolasruscher replied:
Done. JavaDoc with detailed explanation was added, number of results check was added.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:
```diff
@@ -147,29 +147,46 @@ public FileUploadSubmissionResource(SubmissionRepository submissionRepository, R
+        fileUploadExercise.setGradingCriteria(gradingCriteria);
+
+        if (resultId > 0) {
+            fileUploadSubmission = fileUploadSubmissionService.findOneWithEagerResultAndAssessorAndFeedback(submissionId);
```
Please also add some inline comments here. It's e.g. not obvious why we need to reload the submission (it was already loaded from the database above)

nicolasruscher replied:
Done

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java:
```diff
@@ -191,10 +192,23 @@ public ModelingSubmissionResource(SubmissionRepository submissionRepository, Res
         }
-        modelingSubmission = modelingSubmissionService.lockAndGetModelingSubmission(submissionId, modelingExercise, correctionRound);
+
+        if (resultId > 0) {
```
We now have quite some duplicated code even if the code is not specific to modeling / file upload / text.
Would it not make sense to extract it into a common method?

nicolasruscher replied:
I looked into the code, for these calls in general, extraction and unifications across the different exercise types could be done. I guess this should be done carefully and well tested. I think a follow-up would be a good idea for that.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:
```diff
@@ -252,12 +253,14 @@ public TextAssessmentResource(AuthorizationCheckService authCheckService, TextAs
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<Participation> retrieveParticipationForSubmission(@PathVariable Long submissionId,
-            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound) {
+            @RequestParam(value = "correction-round", defaultValue = "0") int correctionRound, @RequestParam(value = "resultId", defaultValue = "0") long resultId) {
```
same feedback as above for modeling and file upload

nicolasruscher replied:
Done for all types

krusche reviewed (changes requested):
I found some code issues and added inline comments. Would be great if you could fix those issues.

nicolasruscher commented:
> I found some code issues and added inline comments. Would be great if you could fix those issues.

I changed the request Param to `Long` with a not required annotation. I added the missing detailed JavaDoc explanation and removed some existing duplications.

FrankeLukas commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -133,6 +133,12 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
                             <span jhiTranslate="artemisApp.studentExamDetail.reviewer">Reviewer</span>
                         </th>
+                        <th>
+                            <span>Has Complaint</span>
```
This needs a translation.

FrankeLukas commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -184,8 +184,8 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
                                 "
-                                >Yes</span
-                            >
+                                jhiTranslate="artemisApp.studentExamDetail.hasComplaintYes"
```
We should probably use `global.generic.yes` here instead.

nicolasruscher replied:
done!

FrankeLukas reviewed (dismissed):
Re-approve since the translation was fixed.

SimonEntholzer reviewed (approved):
The code is fine, I guess the result handling could be improved in a future PR (as it is a bit messy) but overall there are some really nice additional features.
Testing on TS1 worked as expected :+1:

iozbeyli commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -133,6 +133,12 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
                             <span jhiTranslate="artemisApp.studentExamDetail.reviewer">Reviewer</span>
                         </th>
+                        <th>
+                            <span jhiTranslate="artemisApp.studentExamDetail.hasComplaint">Action</span>
```
Is this Action as well?

nicolasruscher replied:
fixed

iozbeyli commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -162,15 +168,78 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
                             </ng-container>
                             <ng-template #noAssessor></ng-template>
                         </td>
+                        <td class="align-middle">
```
I think you shouldn't add td as well if you are not going to show anything.

nicolasruscher replied:
can you clarify that? because I am showing something there.

iozbeyli commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -162,15 +168,78 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
+                            <span
+                                *ngIf="
+                                    exercise?.studentParticipations?.length > 0 &&
+                                    (exercise?.studentParticipations)![0].results?.length > 0 &&
```
You do this check a lot so it might be better to move this to a separate component that definitely has and exercise and results etc. It will cause problems with strict templates as well.

nicolasruscher replied:
Great catch, done!

iozbeyli commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.ts:
```diff
@@ -118,8 +124,13 @@ export class StudentExamDetailComponent implements OnInit {
-                exercise.studentParticipations[0].results.length > 0
+                exercise.studentParticipations[0].results!.length > 0
             ) {
+                if (exercise!.studentParticipations[0].submissions && exercise!.studentParticipations[0].submissions!.length > 0) {
+                    exercise!.studentParticipations[0].submissions![0].results! = exercise.studentParticipations[0].results;
+                    setLatestSubmissionResult(exercise?.studentParticipations[0].submissions?.[0], getLatestSubmissionResult(exercise?.studentParticipations[0].submissions?.[0]));
```
Putting submissions into a variable might reduce the repetition here

nicolasruscher replied:
yes but then we lose the exercise relation to the submissions. This could be solves via a map, but this would mean a function call of get() to retrieve any value, and additionally for quiz exercises are handled differently..

nicolasruscher replied:
Done via new component!

nicolasruscher commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:
```diff
@@ -133,6 +133,12 @@ <h5><span jhiTranslate="artemisApp.studentExamDetail.exercises">Exercises</span>
                             <span jhiTranslate="artemisApp.studentExamDetail.reviewer">Reviewer</span>
                         </th>
+                        <th>
+                            <span jhiTranslate="artemisApp.studentExamDetail.hasComplaint">Action</span>
```
```suggestion
                            <span jhiTranslate="artemisApp.studentExamDetail.hasComplaint">Has Complaint</span>
```

FrankeLukas commented on src/main/webapp/app/exam/manage/student-exams/student-exam-detail-table-row/student-exam-detail-table-row.component.ts:
```diff
@@ -0,0 +1,89 @@
+     * Get an icon for the type of the given exercise.
+     * @param exercise {Exercise}
+     */
+    exerciseIcon(exercise: Exercise): string {
```
We should think about re-using the method of the exercise model:
https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/entities/exercise.model.ts#L146-L158

nicolasruscher replied:
done

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:
```diff
@@ -259,6 +259,18 @@ and not exists (select prs from participation.results prs
+            AND p.student.id = :#{#studentId}
+            AND p.exercise in :#{#exercises}
+            """)
+    List<StudentParticipation> findByStudentIdAndIndividualExercisesWithEagerSubmissionsResultAndAssessorIgnoreTestRuns(@Param("studentId") Long studentId,
```
This is missing a javadoc comment I suppose.

FrankeLukas commented on src/test/javascript/spec/component/exam/manage/student-exams/student-exam-detail-table-row.component.spec.ts:
```diff
@@ -0,0 +1,129 @@
+import { StudentExamDetailComponent } from 'app/exam/manage/student-exams/student-exam-detail.component';
```
`'StudentExamDetailComponent' is defined but never used`

nicolasruscher replied:
Done

FrankeLukas commented on src/test/javascript/spec/component/exam/manage/student-exams/student-exam-detail-table-row.component.spec.ts:
```diff
@@ -0,0 +1,129 @@
+import { Course } from 'app/entities/course.model';
+import { ArtemisDataTableModule } from 'app/shared/data-table/data-table.module';
+import { AlertComponent } from 'app/shared/alert/alert.component';
+import { ArtemisDurationFromSecondsPipe } from 'app/shared/pipes/artemis-duration-from-seconds.pipe';
```
`'ArtemisDurationFromSecondsPipe' is defined but never used`

nicolasruscher replied:
Done

FrankeLukas commented on src/test/javascript/spec/component/exam/manage/student-exams/student-exam-detail-table-row.component.spec.ts:
```diff
@@ -0,0 +1,129 @@
+import { ArtemisDataTableModule } from 'app/shared/data-table/data-table.module';
+import { AlertComponent } from 'app/shared/alert/alert.component';
+import { ArtemisDurationFromSecondsPipe } from 'app/shared/pipes/artemis-duration-from-seconds.pipe';
+import { ArtemisDatePipe } from 'app/shared/pipes/artemis-date.pipe';
```
`'ArtemisDatePipe' is defined but never used`

nicolasruscher replied:
Done

FrankeLukas commented on src/test/javascript/spec/component/exam/manage/student-exams/student-exam-detail.component.spec.ts:
```diff
@@ -24,18 +24,16 @@ import * as sinonChai from 'sinon-chai';
 import { ReactiveFormsModule } from '@angular/forms';
 import { MockTranslateValuesDirective } from '../../../course/course-scores/course-scores.component.spec';
-import { Exercise } from 'app/entities/exercise.model';
+import { Exercise, ExerciseType } from 'app/entities/exercise.model';
```
`'ExerciseType' is defined but never used`

nicolasruscher replied:
Done

FrankeLukas reviewed (dismissed):
Looks good to me otherwise and works. I have a few comments which are not really critical though, so you can count this as approval.

FuchsDominik reviewed (dismissed):
Tested in on ts3 and the steps still work like advertised.

ivanchimeno reviewed (approved):
reapproving

SimonEntholzer reviewed (approved):
Approve code-quality approvements :+1:

FrankeLukas reviewed (approved):
Re-approve
